app/controllers/concerns/authenticates_with_two_factor.rb

```diff
@@ -53,16 +56,50 @@ def authenticate_with_two_factor
 
   private def authenticate_with_two_factor_via_otp(user)
     if valid_otp_attempt?(user) || valid_backup_code_attempt?(user)
-      # Remove any lingering user data from login
-      session.delete(:otp_user_id)
+      # add 2fa device if true
+      browser = Browser.new(request.user_agent)
+      # foce a device name, even if none provided
+      device_name = user_params[:device_name].presence || "#{browser.name} #{browser.platform.name}"
 
-      user.save!
-      sign_in(user, message: :two_factor_authenticated, event: :authentication)
+      add_2fa_device(user, device_name) if user_params[:remember_device] == 'true' && bypass_2fa_enabled?
+      two_factor_successful(user)
     else
       user.increment_failed_attempts
       Rails.logger.info("Failed Login: user=#{user.email} ip=#{request.remote_ip} method=OTP")
       flash.now[:alert] = _('Invalid two-factor code.')
       prompt_for_two_factor(user)
     end
   end
+
+  private def add_2fa_device(user, name)
+    uuid = SecureRandom.uuid
+    expiration = TwoFactorsMemorizedDevice.expiration_timestamp
+
+    user.two_factors_memorized_devices.create!(
+      uuid: uuid,
+      name: name,
+      session_id: session.id,
+      log_in_ip: request.remote_ip,
+      expires_at: expiration,
+    )
+
+    # set cookie and add to two_factors_memorized_devices list
+    cookies.encrypted[:memorized_device] = { value: uuid, expires: expiration }
+  end
+
+  private def using_memorized_device?(user)
+    cookie_uuid = cookies.encrypted[:memorized_device]
+    return false unless cookie_uuid
+
+    # find if cookie exist in active
+    user.two_factors_memorized_devices.active.exists?(uuid: cookie_uuid)
+  end
+
+  private def two_factor_successful(user)
+    # Remove any lingering user data from login
+    session.delete(:otp_user_id)
+
+    user.save!
```

Comment:
Do we remember why this save is here.. i dont see where the user instance is modified

Reply:
@eanders I'm not sure myself. I only copy-pasted the duplicated code section into a separate function.

Reply:
I don't, but I think it might have to do with ensuring that the user account is valid.